Add Whirlpool washer and dryer to Whirlpool integration

mkmer opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Addition of Sensor platform for the washer/dryer appliances.  
State, time_remaining, and dispense level sensors are created under a device for each washer/dryer found in the users account.
Existing Sixth Sense climate now supports device.
Bump whirlpool-sixth-sense to 0.18: https://github.com/abmantis/whirlpool-sixth-sense/compare/0.17.1...0.18
Depends on https://github.com/home-assistant/core/pull/77237

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [x] Dependency upgrade: https://github.com/abmantis/whirlpool-sixth-sense/compare/0.17.1...0.18
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: [#24107](https://github.com/home-assistant/home-assistant.io/pull/24107)

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

mkmer committed:
Added Whirlpool Washer Sensors

mkmer committed:
Added Whirlpool Washer Sensors

mkmer committed:
Added Whirlpool Washer Sensors

mkmer committed:
Removed unused variables

mkmer committed:
Formatted time

mkmer committed:
Time format to hour

mkmer committed:
unused constants

mkmer committed:
Fixed time left format

mkmer committed:
Tests

mkmer committed:
Tests

mkmer committed:
In step for tests

mkmer committed:
In step for tests

mkmer committed:
Merge branch 'home-assistant:dev' into Whirlpool-washer

mkmer committed:
Finished up tests

mkmer committed:
Removed commented code

mkmer committed:
Added No appliance setup_error

mkmer committed:
Removed "0" time test, invalid assumption.
added available property

mkmer committed:
Merge branch 'dev' into Whirlpool-washer

abmantis commented:
Lets keep this as draft before the other region PR gets in to avoid having reviewers reviewing twice the same thing.

mkmer commented:
A draft it is.

mkmer committed:
Merge branch 'dev' into Whirlpool-washer

mkmer committed:
Added platform list: Only unload platforms loaded

mkmer committed:
Merge branch 'Whirlpool-washer' of https://github.com/mkmer/core into Whirlpool-washer

mkmer committed:
Time remaining device class setup

mkmer committed:
Test cleanup

mkmer committed:
added test for dryer and washer

mkmer committed:
Merge branch 'home-assistant:dev' into Whirlpool-washer

mkmer committed:
Merge branch 'dev' into Whirlpool-washer

mkmer committed:
Merge branch 'dev' into Whirlpool-washer

mkmer committed:
Added Washer cycle states

mkmer committed:
Merge branch 'Whirlpool-washer' of https://github.com/mkmer/core into Whirlpool-washer

mkmer committed:
Merge branch 'home-assistant:dev' into Whirlpool-washer

mkmer committed:
Merge branch 'home-assistant:dev' into Whirlpool-washer

mkmer committed:
Added Door status

mkmer committed:
Fixed door status test

mkmer committed:
Merge branch 'dev' into Whirlpool-washer

mkmer committed:
eliminate mutliple eventsocket connections per appliance

mkmer committed:
Merge branch 'Whirlpool-washer' of https://github.com/mkmer/core into Whirlpool-washer

mkmer committed:
bump to 0.18.0
address new callback registry in climate.py

mkmer committed:
Merge branch 'dev' into Whirlpool-washer

mkmer committed:
Added Maytag supported brand

mkmer committed:
Added Maytag supported brand

mkmer committed:
Merge branch 'dev' into Whirlpool-washer

mkmer committed:
Clean up cycle state lookup

mkmer committed:
Merge branch 'home-assistant:dev' into Whirlpool-washer

mkmer committed:
Address missing import

mkmer committed:
Merge branch 'Whirlpool-washer' of https://github.com/mkmer/core into Whirlpool-washer

mkmer committed:
Fixed Sensor init tab error
Updated tests for new API

mkmer committed:
Finished tests for sensor

mkmer committed:
Merge branch 'dev' into Whirlpool-washer

mkmer committed:
Added Door Open test

mkmer committed:
Cleaned up platform loading

mkmer committed:
Merge branch 'home-assistant:dev' into Whirlpool-washer

mkmer committed:
Back to Unbound call list

mkmer committed:
Merge branch 'Whirlpool-washer' of https://github.com/mkmer/core into Whirlpool-washer

mkmer committed:
Removed incorrect model in device_info

mkmer committed:
Updated register callback function name

mkmer committed:
Merge branch 'dev' into Whirlpool-washer

mkmer committed:
Merge branch 'dev' into Whirlpool-washer

mkmer committed:
hassfest updates

mkmer committed:
hassfest updates

mkmer committed:
New function names missed in tests

mkmer committed:
Merge branch 'dev' into Whirlpool-washer

mkmer committed:
Merge branch 'home-assistant:dev' into Whirlpool-washer

mkmer committed:
Merge branch 'home-assistant:dev' into Whirlpool-washer

mkmer committed:
Updated doc strings.

mkmer committed:
Merge branch 'dev' into Whirlpool-washer

mkmer committed:
Merge branch 'dev' into Whirlpool-washer

mkmer committed:
Remove commented code

mkmer committed:
Merge branch 'dev' into Whirlpool-washer

mkmer committed:
Merge branch 'home-assistant:dev' into Whirlpool-washer

mkmer committed:
Merge branch 'dev' into Whirlpool-washer

mkmer committed:
Merge branch 'dev' into Whirlpool-washer

mkmer committed:
Merge branch 'dev' into Whirlpool-washer

mkmer committed:
Merge branch 'dev' into Whirlpool-washer

mkmer committed:
Merge branch 'dev' into Whirlpool-washer

mkmer committed:
Merge branch 'dev' into Whirlpool-washer

mkmer committed:
Multiple devices now supported = HUB

mkmer committed:
Close socket on removal.
Eventsocket was left running on reload

mkmer committed:
Merge branch 'dev' into Whirlpool-washer

mkmer committed:
Address Async Mock for disconnect

mkmer committed:
Remove unused vars
Finish tests for sensor.py

mkmer committed:
Merge branch 'dev' into Whirlpool-washer

mkmer committed:
Merge branch 'dev' into Whirlpool-washer

mkmer committed:
Merge branch 'dev' into Whirlpool-washer

mkmer committed:
Merge branch 'home-assistant:dev' into Whirlpool-washer

mkmer committed:
Catch timeout error in api

mkmer committed:
ConfigAuthFailed added

mkmer committed:
Merge branch 'dev' into Whirlpool-washer

mkmer committed:
Merge branch 'dev' into Whirlpool-washer

mkmer committed:
Merge branch 'dev' into Whirlpool-washer

mkmer committed:
Merge branch 'home-assistant:dev' into Whirlpool-washer

mkmer committed:
Removed configauth - move to alternate PR

mkmer committed:
Merge branch 'dev' into Whirlpool-washer

mkmer committed:
More generic user input for region (and more)
Add region to tests.

mkmer committed:
Merge branch 'dev' into Whirlpool-washer

mkmer committed:
Merge branch 'dev' into Whirlpool-washer

mkmer committed:
Added Dispense Tank level sensor

mkmer committed:
Fix tank fill values

mkmer committed:
Add dispense level test

mkmer committed:
Move constants to const.py

mkmer committed:
Merge branch 'dev' into Whirlpool-washer

mkmer committed:
Add additional fill status

mkmer committed:
Add default unknown for state lookup

mkmer committed:
Merge branch 'dev' into Whirlpool-washer

bramkragten committed:
Update frontend to 20221213.1 (#84058)

epenet committed:
Add device classes to bloomsky (#84054)

bdraco committed:
Update ESPHome bluetooth repair issue for unique id change and 2022.12.0 (#84026)

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

DataBitz committed:
Fix grammatical error in HomeKit strings (#84043)

Co-authored-by: J. Nick Koston <nick@koston.org>

epenet committed:
Add device classes in tesla wall connector (#84048)

Addd device classes in tesla wall connector

GitHub-Action committed:
[ci skip] Translation update

epenet committed:
Move unused function in withings (#84089)

* Remove unused function in withings

* Move to tests

* Remove reference to attribute.platform

* Move WITHINGS_MEASUREMENTS_MAP to tests

* measute_type > measure_type

* One more

epenet committed:
Refactor withings constant (#84095)

Split withings constant

frenck committed:
Rename image integration to image_upload (#84063)

* Rename image integration to image_upload

* fix test

elupus committed:
Make sure philips_hs remote entity calls parent added (#84082)

Make sure remote calls parent added

bdraco committed:
Bump bluetooth-auto-recovery to 1.0.3 (#84075)

* Bump bluetooth-auto-recovery to 1.0.1

changelog: https://github.com/Bluetooth-Devices/bluetooth-auto-recovery/compare/v1.0.0...v1.0.1

Handles the case were the adapter gets a new hci number
after a usb reset.

* one more

* bump again

bdraco committed:
Bump govee-ble to 0.19.3 (#84062)

* Bump govee-ble to 0.19.2

fixes #82994 fixes #81335

changelog: https://github.com/Bluetooth-Devices/govee-ble/compare/v0.19.1...v0.19.2

* .3

muppet3000 committed:
Re-add missing Growatt TLX values (#84040)

* Growatt - Re-adding missing TLX values (#81470)

* Growatt - Added missing files to .coveragerc

* Growatt - Correcting TLX Voltage types

* Growatt - Addressing review comments

epenet committed:
Split platform attributes in withings (#84099)

DCSBL committed:
Improve HomeWizard request issue reporting (#82366)

* Trigger reauth flow when HomeWizard API was disabled

* Add tests for reauth flow

* Fix typo in test

* Add parallel updates constant

* Improve error message when device in unreachable during config

* Set quality scale

* Remove quality scale

* Throw error instead of abort when setup fails

* Adjust test for new setup behaviour

* Trigger reauth flow when API is disabled and continue retrying

* Reload entry and raise AuthFailed during init

* Abort running config flow

* Listen for coordinator updates to trigger reload

* Use build-in backoff system

* Fix failing test

* Test reauth flow is active after disable-api init

* Test reauth flow removal

epenet committed:
Add device class to withings (#84103)

Danielhiversen committed:
Update Tibber lib, improve realtime streaming (#84065)

Update tibber lib, improve realtime streaming

Signed-off-by: Daniel Hjelseth Høyer <github@dahoiv.net>

Signed-off-by: Daniel Hjelseth Høyer <github@dahoiv.net>

GitHub-Action committed:
[ci skip] Translation update

cdce8p committed:
Fix prettier [homewizard] (#84122)

bdraco committed:
Rename Bluetooth active_update_coordinator to active_update_processor (#84115)

epenet committed:
Add device classes in starline (#84049)

epenet committed:
Add device classes to pool integrations (#84050)

* Add device classes to pool integrations

* Update omnilogic

* Don't use VOLTAGE device class

bieniu committed:
Bump `brother` and `pysnmplib` (#84107)

* Bump brother version

* Bump pysnmplib version

* Update sensor platform

* Update switch platform

* Update tests

* Bump brother

Co-authored-by: J. Nick Koston <nick@koston.org>

nyroDev committed:
Fix current_operation when there is no operation_mode in Overkiz DHWP (#84077)

Fix current_operation when there is no operation_mode

bonzini committed:
Make manual and manual_mqtt fully asynchronous (#83436)

* manual: make it completely async

Restoring the timers on startup cannot use track_point_in_time,
because the restoring code runs from the async_added_to_hass
method.  Rewrite everything to run in the event loop instead
of using threaded wrappers, this way the code can be reused in
async_added_to_hass.

* manual_mqtt: replace async function with @callback

* manual_mqtt: make it completely async

Restoring the timers on startup cannot use track_point_in_time,
because the restoring code runs from the async_added_to_hass
method.  Rewrite everything to run in the event loop instead
of using threaded wrappers, this way the code can be reused in
async_added_to_hass when state restore is added.

Co-authored-by: Jan Bouwhuis <jbouwh@users.noreply.github.com>

cdce8p committed:
Update pyupgrade to 3.3.1 (#84148)

cdce8p committed:
Update isort to 5.11.3 (#84147)

tronikos committed:
Google Assistant SDK: Support non en-US language code (#84028)

* Support non en-US language code

* Get default language_code based on HA config

* Revert bumping gassist-text
Will be done in a separate PR

MartinHjelmare committed:
Clean up matter adapter (#84144)

farmio committed:
Fix KNX ConfigFlow for manual secure tunnel keys (#84155)

GitHub-Action committed:
[ci skip] Translation update

allenporter committed:
Bump ical to 4.2.3 (#84104)

mib1185 committed:
Fix check if Surveillance Station is available in Synology DSM (#84140)

start coordinator only when surveillance_station is avialble during api setup

bdraco committed:
Add dhcp discovery for newer august/yale bridges (#84125)

bdraco committed:
Bump yalexs-ble to 1.12.2 (#84118)

eifinger committed:
Enable strict typing for here_travel_time (#83814)

* Enable strict typing for here_travel_time

* Use dict[str, Any] instead of Any

* Use kwargs ctor

* Specify dicts

engrbm87 committed:
Move speedtestdotnet coordinator to separate file (#83979)

J3173 committed:
Add state_class to Tradfri battery and air quality sensors (#84185)

albertogeniola committed:
Elmax/alarm_control_panel (#75365)

Ernst79 committed:
Sensor pro fix for negative temperatures (#84181)

fixes undefined

bachya committed:
Add re-auth flow to AirVisual Pro (#84012)

* Add re-auth flow to AirVisual Pro

* Code review

bachya committed:
Save entry earlier in PurpleAir re-auth flow (#84203)

bachya committed:
Add ability to add/remove PurpleAir sensors in an existing config entry (#83440)

hahn-th committed:
Add device HmIP-DRDI3 (#83337)

epenet committed:
Use DATA_SIZE device class in folder integration (#83897)

GitHub-Action committed:
[ci skip] Translation update

bdraco committed:
Bump PySwitchbot to 0.29.1 (#84119)

fixes https://github.com/home-assistant/core/issues/82795
closes https://github.com/home-assistant/core/issues/78562

Djelibeybi committed:
Bump aiolifx-themes to 0.4.0 (#84216)

epenet committed:
Use new unit enumerators in tibber (#84222)

epenet committed:
Use UnitOfEnergy in integrations (o-z) (#84221)

epenet committed:
Use UnitOfEnergy in integrations (a-n) (#84220)

bieniu committed:
Bump `nettigo air monitor` library to version 1.6.0 (#84129)

bonzini committed:
Fix manual.alarm_control_panel RestoreEntity bugs (#82990)

* manual: add previous state for TRIGGERED as well

This will be useful to properly restore timers on restart.

* manual: correctly restore timers on startup

epenet committed:
Use ENERGY device class in landisgyr heat meter (#84223)

epenet committed:
Use ENERGY device class in oru (#84225)

* Use ENERGY device class in oru

* Apply suggestions from code review

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

bdraco committed:
Bump zeroconf to 0.44.0 (#84166)

epenet committed:
Use new unit enums in fronius (#84224)

epenet committed:
Use UnitOfVolume in bmw_connected_drive (#84228)

michaelarnauts committed:
Use a global const for CONF_UUID (#84093)

tronikos committed:
Bump gassist-text to 0.0.5 (#84161)

boralyl committed:
Upgrade attrs to 22.1.0 (#84205)

MartinHjelmare committed:
Clean mysensors test type errors (#84178)

epenet committed:
Use new unit enumerators in tuya (#84229)

epenet committed:
Use UnitOfVolume in dsmr_reader (#84226)

epenet committed:
Add POWER device class to goodwe number (#84232)

epenet committed:
Adjust units in lcn (#84051)

* Use unit enumerators in lcn

* Don't mix lcn units with home assistant units

bdraco committed:
Fix connectable Bluetooth devices not going available after scanner recovers (#84172)

engrbm87 committed:
Remove name key from transimssion services (#83974)

frenck committed:
Update sentry-sdk to 1.12.1 (#84234)

ollo69 committed:
Add integration-type property to Nut manifest (#84020)

epenet committed:
Add support for CCF volume unit (#84029)

frenck committed:
Update coverage to 7.0.0 (#84237)

frenck committed:
Update black to 22.12.0 (#84242)

frenck committed:
Update spotipy to 2.22.0 (#84244)

epenet committed:
Use UnitOfElectricCurrent in integrations (#84231)

frenck committed:
Update watchdog to 2.2.0 (#84245)

bdraco committed:
Use a single source of data for switchbot (#84215)

bachya committed:
Migrate AirVisual Pro devices to the `airvisual_pro` domain (#83882)

* Migration AirVisual Pro devices to the `airvisual_pro` domain

* Fix tests

* Remove airvisual_pro dependency

* Add repairs item

* Only fire repairs issue if automations exist

* Fix tests

allenporter committed:
Bump ical to 4.2.4 (#84248)

mkmer committed:
Default test

mkmer committed:
clean up post rebase

mkmer commented:
Branch messed up.. creating new PR
